Mathlib/RingTheory/Polynomial/DegreeLT.lean

```diff
@@ -0,0 +1,203 @@
+  rw [← (basis ..).sum_repr P]
+  simp [-LinearMap.coe_inl, addLinearEquiv_symm_apply_inl_xpow]
+
+lemma addLinearEquiv_symm_apply_inr (Q : R[X]_n) :
+    ((addLinearEquiv R m n).symm (LinearMap.inr R _ _ Q) : R[X]) = (Q : R[X]) * X ^ (m : ℕ) := by
+  rw [← (basis ..).sum_repr Q]
+  simp [-LinearMap.coe_inr, Finset.sum_mul, addLinearEquiv_symm_apply_inr_xpow,
+    smul_eq_C_mul, mul_assoc, ← pow_add, add_comm]
+
+lemma addLinearEquiv_symm_apply (PQ) :
+    ((addLinearEquiv R m n).symm PQ : R[X]) = (PQ.1 : R[X]) + (PQ.2 : R[X]) * X ^ (m : ℕ) := calc
+  _ = ((addLinearEquiv R m n).symm (LinearMap.inl R _ _ PQ.1 + LinearMap.inr R _ _ PQ.2) : R[X]) :=
+      by rw [LinearMap.inl_apply, LinearMap.inr_apply, Prod.add_def, add_zero, zero_add]
+  _ = _ := by rw [map_add, Submodule.coe_add,
+        addLinearEquiv_symm_apply_inl, addLinearEquiv_symm_apply_inr]
+
+lemma addLinearEquiv_symm_apply' (PQ) :
+    ((addLinearEquiv R m n).symm PQ : R[X]) = (PQ.1 : R[X]) + X ^ (m : ℕ) * (PQ.2 : R[X]) := by
+  rw [X_pow_mul, addLinearEquiv_symm_apply]
+
+lemma addLinearEquiv_apply' {R : Type*} [Ring R] (f) :
+    ((addLinearEquiv R m n f).1 : R[X]) = f %ₘ (X ^ m) ∧
+      ((addLinearEquiv R m n f).2 : R[X]) = f /ₘ (X ^ m) := by
+  rw [and_comm, eq_comm, eq_comm (b := _ %ₘ _)]
+  nontriviality R; refine div_modByMonic_unique _ _ (monic_X_pow _) ⟨?_, ?_⟩
+  · rw [← addLinearEquiv_symm_apply', LinearEquiv.symm_apply_apply]
+  · rw [degree_X_pow, ← mem_degreeLT]; exact Subtype.prop _
+
+lemma addLinearEquiv_apply_fst {R : Type*} [Ring R] (f) :
+    ((addLinearEquiv R m n f).1 : R[X]) = f %ₘ (X ^ m) :=
+  (addLinearEquiv_apply' f).1
+
+lemma addLinearEquiv_apply_snd {R : Type*} [Ring R] (f) :
+    ((addLinearEquiv R m n f).2 : R[X]) = f /ₘ (X ^ m) :=
+  (addLinearEquiv_apply' f).2
+
+lemma addLinearEquiv_apply {R : Type*} [Ring R] (f) :
+    addLinearEquiv R m n f =
+      (⟨f %ₘ (X ^ m), addLinearEquiv_apply_fst f ▸ Subtype.prop _⟩,
+      ⟨f /ₘ (X ^ m), addLinearEquiv_apply_snd f ▸ Subtype.prop _⟩) :=
+  Prod.ext (Subtype.ext <| addLinearEquiv_apply_fst f) (Subtype.ext <| addLinearEquiv_apply_snd f)
+
+end degreeLT
+
+section taylor
+
+variable {R : Type*} [CommRing R] {r : R} {m n : ℕ} {s : R} {f g : R[X]}
+
+lemma map_taylorEquiv_eq_degreeLT : (R[X]_n).map (taylorEquiv r) = R[X]_n :=
+  le_antisymm (Submodule.map_le_iff_le_comap.2
+      fun _ hf ↦ mem_degreeLT.2 <| (degree_taylor ..).trans_lt <| mem_degreeLT.1 hf)
+    (fun f hf ↦ (taylorEquiv r).apply_symm_apply f ▸ Submodule.mem_map_of_mem
+      (mem_degreeLT.2 <| (degree_taylor ..).trans_lt <| mem_degreeLT.1 hf))
+
+/-- The map `taylor r` induces an automorphism of the module `R[X]_n` of polynomials of
+degree `< n`. -/
+noncomputable def taylorLinear (r : R) (n : ℕ) : R[X]_n ≃ₗ[R] R[X]_n :=
+  (taylorEquiv r : R[X] ≃ₗ[R] R[X]).ofSubmodules _ _ map_taylorEquiv_eq_degreeLT
+
+@[simp] lemma taylorLinear_apply (P : R[X]_n) :
+    (taylorLinear r n P : R[X]) = (P : R[X]).taylor r :=
+  rfl
```

Comment:
Can you generate this one with `@[simps]` on `taylorLinear`?

Reply:
Actually, since `@[simps!]` generates the incorrect lemma for the backwards version, I think it is actually better to not have `@[simps!]` declared. What do you think?

Reply:
You can do `@[simps! apply]` (with `apply` replaced by the actual projection it generates, I expect it'll actually be `apply_val`) to only generate the forward direction

Reply:
`@[simps! apply]` is the correct call (it does not support using the actual name it generates), and it gives:
```lean4
Polynomial.taylorLinearEquiv_apply.{u_1} {R : Type u_1} [CommRing R] (r : R) (n : ℕ) (a✝ : ↥R[X]_n) :
  (taylorLinearEquiv r n) a✝ =
    (LinearEquiv.ofEq (Submodule.map (↑↑(taylorEquiv r)) R[X]_n) (R[X]_n) ⋯)
      (((↑(taylorEquiv r)).submoduleMap R[X]_n) a✝)
```
which is technically a correct step in the tree but I feel like I would rather not have this. What do you think?

Reply:
`apply_coe` is the projection you want. I've pushed a fix